File: src/leetcode/p279/Solution.java
package leetcode.p279;
// https://leetcode.com/problems/perfect-squares/

public class Solution {
    private int[] cache;

    public int numSquares(int n) {
        cache= new int[n+1];
        for (int i=0; i <= n; ++i) cache[i]= -1;
        return minNumOfSquares(n);
    }

    private int minNumOfSquares(int num) {
        if (num <= 1) return num;
        if (cache[num] > -1) return cache[num];

        cache[num]= num;
        for (int i=1; i*i <= num; ++i) {
            cache[num]= Math.min(cache[num], minNumOfSquares(num - i*i) + 1);
        }
        return cache[num];
    }
}

File: src/leetcode/go/p279/solution.go
/*
Leetcode problem p279 solution.

Dynamic Programming (Top Down)
- Time: O(n sqrt(n)^2)
- Space: O(n)
*/
package p279

var cc [10001]int

func numSquares(n int) int {
	for i:=0; i <= 10000; i++ {
		cc[i]= -1
	}
	return getMinCount(n)
}

func getMinCount(num int) int {
	if cc[num] > -1 {
		return cc[num]
	}

	cc[num]= num
	for i:=1; i*i <= num; i++ {
		cc[num] = min(cc[num], 1 + getMinCount(num-i*i))
	}
	return cc[num]
}

